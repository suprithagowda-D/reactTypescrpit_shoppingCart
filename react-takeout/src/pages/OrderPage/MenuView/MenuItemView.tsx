import { useOrderContext } from '../OrderContext';
import { useToasts } from 'react-toast-notifications';
import { MenuItem } from '../../../models/Menu';

export interface MenuItemViewProps {
  menuItem: MenuItem;
}

export const MenuItemView = ({ menuItem }: MenuItemViewProps) => {
  const { name, description, price } = menuItem;
  const { order, setOrder } = useOrderContext();
  const { addToast } = useToasts();

  const handleItemClicked = (productId: string) => {
    const items = order.items.some((menu) => menu.menuItemId === productId);
    if (!items) {
      setOrder({
        ...order,
        items: [...order.items, { menuItemId: productId, name, price }],
      });
    } else {
      addToast('Item already in cart', {
        appearance: 'info',
        autoDismiss: true,
      });
    }
  };
  return (
    <div onClick={() => handleItemClicked(menuItem.id)}>
      <li className={`mb-2 cursor-pointer`}>
        <h4>{name}</h4>
        <p>{description}</p>
        <h5>{price}</h5>
      </li>
    </div>
  );
};
