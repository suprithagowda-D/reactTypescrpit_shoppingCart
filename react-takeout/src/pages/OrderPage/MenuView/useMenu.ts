import { useState, useEffect } from 'react';
import { EnvVar } from '../../../utils';
import { useEnv } from '../../../contexts';
import { Menu } from '../../../models';

export const useMenu = () => {
  const env = useEnv();
  const apiUrl = env.get(EnvVar.API_URL);

  const [isLoading, setIsLoading] = useState(false);
  const [isError, setIsError] = useState(false);
  const [error, setError] = useState<unknown>();
  const [menu, setMenu] = useState<Menu>();

  useEffect(() => {
    const fetchMenus = async () => {
      try {
        setIsLoading(true);
        const response = await fetch(`${apiUrl}/menu`);

        if (!response.ok) {
          const message = `Error: ${response.status}`;
          throw new Error(message);
        }

        const menus = await response.json();
        setMenu(menus);
        setIsLoading(false);
      } catch (error) {
        setIsError(true);
        setError(error);
        setIsLoading(false);
      }
    };

    fetchMenus();
  }, [apiUrl]);

  return { isLoading, isError, error, menu };
};
