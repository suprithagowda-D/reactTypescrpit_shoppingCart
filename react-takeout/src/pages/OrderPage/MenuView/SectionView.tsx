import { Menu } from '../../../models/Menu';
import { MenuItemView } from './MenuItemView';

export interface MenuListProps {
  menu: Menu;
}

export const SectionView = ({ menu }: MenuListProps) => {
  const menuList = menu.sectionIds.map((items, key) => {
    const itemId = menu.sections[items].itemIds;
    const itemList = itemId.map((itemName) => {
      const menuItem = menu.items[itemName];
      return <MenuItemView key={menuItem.id} menuItem={menuItem} />;
    });
    return (
      <div>
        <h3 className="title2">{items} </h3>
        <ul>{itemList}</ul>
      </div>
    );
  });
  return <>{menuList}</>;
};
