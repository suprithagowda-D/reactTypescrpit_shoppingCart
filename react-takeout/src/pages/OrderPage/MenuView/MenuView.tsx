import { useMenu } from './useMenu';
import { Loading } from '../../../components';
import { StringUtils } from '../../../utils';
import { SectionView } from './SectionView';

export const MenuView = () => {
  const { isLoading, isError, error, menu } = useMenu();

  if (isLoading) {
    return <Loading />;
  }

  if (isError) {
    return (
      <main>
        <h1>{StringUtils.errorToString(error)}</h1>
      </main>
    );
  }

  return (
    <div className={`overflow-auto min-h-0 divider-right width60`}>
      <h1 className="title">Our Menu</h1>
      <main>{menu && <SectionView menu={menu} />}</main>
    </div>
  );
};
