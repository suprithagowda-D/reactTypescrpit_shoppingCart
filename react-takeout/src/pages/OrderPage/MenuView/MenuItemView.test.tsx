import React from 'react';
import { render } from '../../../test/test-utils';
import { MenuView } from './MenuView';

describe('<MenuItemView />', () => {
  test('renders correctly', () => {
    const { asFragment } = render(<MenuView />);
    expect(asFragment()).toMatchSnapshot();
  });
});
