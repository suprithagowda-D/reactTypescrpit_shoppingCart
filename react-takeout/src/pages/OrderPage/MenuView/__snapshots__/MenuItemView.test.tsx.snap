// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuItemView /> renders correctly 1`] = `
<DocumentFragment>
  <div>
    Loading...
  </div>
</DocumentFragment>
`;
